overlord/devicestate/handlers.go

```diff
@@ -698,3 +703,231 @@ func fetchKeys(st *state.State, keyID string) (errAcctKey error, err error) {
 		keyID = a.SignKeyID()
 	}
 }
+
+func snapState(st *state.State, name string) (*snapstate.SnapState, error) {
+	var snapst snapstate.SnapState
+	err := snapstate.Get(st, name, &snapst)
+	if err != nil && err != state.ErrNoState {
+		return nil, err
+	}
+	return &snapst, nil
+}
+
+func makeRollbackDir(name string) (string, error) {
+	snapRollbackDir := filepath.Join(dirs.SnapRollbackDir, name)
+
+	if err := os.MkdirAll(dirs.SnapRollbackDir, 0755); err != nil {
+		return "", err
+	}
+
+	if err := os.Mkdir(snapRollbackDir, 0755); err != nil {
+		// fail even if directory exists already
+		return "", err
+	}
+
+	return snapRollbackDir, nil
+}
+
+func currentGadgetInfo(snapst *snapstate.SnapState) (*gadget.Info, error) {
+	var gi *gadget.Info
+
+	currentInfo, err := snapst.CurrentInfo()
+	if err != nil && err != snapstate.ErrNoCurrent {
+		return nil, err
+	}
+	if currentInfo != nil {
+		gi, err = snap.ReadGadgetInfo(currentInfo, release.OnClassic)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return gi, nil
+}
+
+func pendingGadgetInfo(snapsup *snapstate.SnapSetup) (*gadget.Info, error) {
+	info, err := snap.ReadInfo(snapsup.InstanceName(), snapsup.SideInfo)
+	if err != nil {
+		return nil, err
+	}
+	update, err := snap.ReadGadgetInfo(info, release.OnClassic)
+	if err != nil {
+		return nil, err
+	}
+	return update, nil
+}
+
+func gadgetCurrentAndUpdate(st *state.State, snapsup *snapstate.SnapSetup) (current *gadget.Info, update *gadget.Info, err error) {
+	snapst, err := snapState(st, snapsup.InstanceName())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	currentInfo, err := currentGadgetInfo(snapst)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	newInfo, err := pendingGadgetInfo(snapsup)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return currentInfo, newInfo, nil
+}
+
+var (
+	gadgetUpdate        = nopGadgetOp
+	gadgetRollback      = nopGadgetOp
+	gadgetTrashRollback = nopGadgetTrashRollback
+)
+
+func nopGadgetOp(current, update *gadget.Info, rollbackRootDir string) error {
+	return nil
+}
+
+func nopGadgetTrashRollback(current *gadget.Info, rollbackRootDir string) error {
+	return nil
+}
+
+func (m *DeviceManager) doUpdateGadget(t *state.Task, _ *tomb.Tomb) error {
+	if release.OnClassic {
```

Comment:
Is it relevant? I think we could just leave it out as we don't expect gadgets to be actually carrying any assets on classic.

Reply:
Discussed with @mvo5 and will drop that.